File: seeker/snippet/quickSort.java
//date: 2022-03-09T16:57:24Z
//url: https://api.github.com/gists/4f23013941f6e0f31d1f47f518200171
//owner: https://api.github.com/users/Danush-TBR

import java.util.Scanner;

public class quickSort {
    public static int partition(int[] arr,int start,int end){
        int i=start, j=end-1, PivotIndex = end;
        while(i<j){
            while(i<end && arr[i]<=arr[PivotIndex]) i++;
            while(j>start && arr[j]>arr[PivotIndex]) j--;
            if(i<j){
                int temp = arr[i];
                arr[i]=arr[j];
                arr[j]=temp;
            }
        }
        int temp=arr[PivotIndex];
        arr[PivotIndex]=arr[i];
        arr[i]=temp;
        return i;
    }
    public static void QuickSort(int[] arr,int start,int end){
        if(start<end){
            int partitionIndex = partition(arr, start, end);
            QuickSort(arr, start, partitionIndex-1);
            QuickSort(arr, partitionIndex+1, end);
        }
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N=scanner.nextInt();
        int[] arr = new int[N];
        for(int i=0;i<N;i++) arr[i]=scanner.nextInt();
        QuickSort(arr,0,N-1);
        for(int i=0;i<N;i++) System.out.print(arr[i]+" ");
        scanner.close();
    }
}

File: seeker/snippet/quick_sort.py
#date: 2022-07-01T17:06:17Z
#url: https://api.github.com/gists/4dc020cf7daa38a0182c8f3fcc0a9875
#owner: https://api.github.com/users/adrolc

'''
Time efficiency test

Algorithm: quicksort. Minimum execution time: 0.10355029999999998
Algorithm: quicksort2. Minimum execution time: 0.17530590000000013
'''

import random

# not in-place
def quicksort(array):
    if len(array) < 2:
        return array

    low, same, high = [], [], []
    pivot = array[random.randint(0, len(array) - 1)]

    for item in array:
        if item < pivot:
            low.append(item)
        elif item == pivot:
            same.append(item)
        elif item > pivot:
            high.append(item)
    return quicksort(low) + same + quicksort(high)


# in-place
def quicksort2(array, low=0, high=None):
    if high == None:
        high = len(array) - 1

    if low >= high:
        return

    i, j = low, high
    pivot = array[random.randint(low, high)]

    while i <= j:
        while array[i] < pivot:
            i += 1
        while array[j] > pivot:
            j -= 1

        if i <= j:
            array[i], array[j] = array[j], array[i]
            i, j = i + 1, j - 1

    quicksort2(array, low, j)
    quicksort2(array, i, high)
